core/controller/block.ts

```diff
@@ -0,0 +1,73 @@
+import Response from 'shared/model/response';
+import { Block } from 'shared/model/block';
+import { BlockService } from 'core/service/block';
+import { BlockRepo } from 'core/repository/block';
+import { Peer } from 'shared/model/peer';
+import { ON, RPC } from "core/util/decorator";
+import { BaseController } from 'core/controller/baseController';
+
+interface BlockGenerateRequest {
+    keypair: {
+        privateKey: string,
+        publicKey: string
+    };
+    timestamp: number;
+}
+
+class BlockController extends BaseController {
+    private blockService: BlockService = new BlockService();
+    private blockRepo: BlockRepo = new BlockRepo();
+
+    @ON('BLOCK_RECEIVE')
+    public async onReceiveBlock(block: Block): Promise<Response<void>> {
+        const response: Response<void> = await this.blockService.processIncomingBlock(block);
+        if (!response.success) {
+            response.errors.push('onReceiveBlock');
+        }
+        return response;
+    }
+
+    @ON('BLOCK_GENERATE')
+    public async generateBlock( data: BlockGenerateRequest ): Promise<Response<void>> {
+        const response: Response<void> = await this.blockService.generateBlock(data.keypair, data.timestamp);
+        if (!response.success) {
+            response.errors.push('generateBlock');
+        }
+        return response;
+    }
+
+    @ON('BLOCKCHAIN_READY')
+    public async loadLastNBlocks(): Promise<Response<void>> {
+        const response: Response<string[]> = await this.blockRepo.loadLastNBlocks();
+        if (!response.success) {
+            return new Response<void>({ errors: [...response.errors, 'loadLastNBlocks'] });
+        }
+        const blocks: string[] = response.data;
+        this.blockService.setLastNBlocks(blocks);
+        return new Response<void>();
+    }
+
+    @ON('NEW_BLOCKS')
+    public updateLastNBlocks(block: Block): Response<void> {
+        this.blockService.updateLastNBlocks(block);
+        return new Response<void>();
+    }
+
+    @RPC('GET_COMMON_BLOCK')
+    // called from UI
```

Comment:
This logic should not be called from RPC

Reply:
Also used for sync between nodes, but called from current UI as well. I'm not sure if this method is necessary here.

Reply:
Method for sync between nodes cannot be RPC or API. It must be private

Reply:
But here can be a method for getting a common block from another nodes. Like a https://github.com/AraiEzzra/DDKCORE/blob/ed086aad0cb545fd952f5ef1da38820c89014200/src/modules/transport.js#L556